File: android/app/src/main/java/com/example/data/IcdService.kt
package com.example.data

import android.util.Base64
import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.FormBody
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject
import java.util.concurrent.TimeUnit

data class IcdSearchResult(
    val code: String,
    val title: String,
    val uri: String
)

object IcdService {
    private const val TAG = "IcdService"
    private const val TOKEN_URL = "https://icdaccessmanagement.who.int/connect/token"
    private const val SEARCH_URL = "https://id.who.int/icd/entity/search"

    var customClientId: String = ""
    var customClientSecret: String = ""

    private var cachedToken: String = ""
    private var tokenExpiryTime: Long = 0

    private val client = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .build()

    private suspend fun getAccessToken(): String? = withContext(Dispatchers.IO) {
        if (customClientId.isEmpty() || customClientSecret.isEmpty()) {
            return@withContext null
        }

        // Return cached token if valid
        val currentTime = System.currentTimeMillis()
        if (cachedToken.isNotEmpty() && currentTime < tokenExpiryTime) {
            return@withContext cachedToken
        }

        try {
            val authHeader = "Basic " + Base64.encodeToString(
                "$customClientId:$customClientSecret".toByteArray(Charsets.UTF_8),
                Base64.NO_WRAP
            )

            val requestBody = FormBody.Builder()
                .add("grant_type", "client_credentials")
                .add("scope", "icdapi_access")
                .build()

            val request = Request.Builder()
                .url(TOKEN_URL)
                .post(requestBody)
                .addHeader("Authorization", authHeader)
                .build()

            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    Log.e(TAG, "Token request failed: ${response.code}")
                    return@withContext null
                }
                val body = response.body?.string() ?: ""
                val json = JSONObject(body)
                val token = json.getString("access_token")
                val expiresIn = json.getLong("expires_in") // seconds

                cachedToken = token
                tokenExpiryTime = System.currentTimeMillis() + (expiresIn - 60) * 1000 // Buffer of 60 seconds
                return@withContext token
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error getting access token", e)
            return@withContext null
        }
    }

    suspend fun searchIcd11(query: String): List<IcdSearchResult> = withContext(Dispatchers.IO) {
        val token = getAccessToken()
        if (token == null) {
            Log.i(TAG, "Using local fallback ICD-11 search for query: $query")
            return@withContext getLocalFallbackResults(query)
        }

        try {
            val url = "$SEARCH_URL?q=${java.net.URLEncoder.encode(query, "UTF-8")}"
            val request = Request.Builder()
                .url(url)
                .addHeader("Authorization", "Bearer $token")
                .addHeader("API-Version", "v2")
                .addHeader("Accept-Language", "en")
                .build()

            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    Log.e(TAG, "Search request failed: ${response.code}. Falling back to local data.")
                    return@withContext getLocalFallbackResults(query)
                }

                val body = response.body?.string() ?: ""
                val json = JSONObject(body)
                val results = mutableListOf<IcdSearchResult>()
                
                val destinationEntities = json.optJSONArray("destinationEntities")
                if (destinationEntities != null) {
                    for (i in 0 until destinationEntities.length()) {
                        val entity = destinationEntities.getJSONObject(i)
                        val titleRaw = entity.optString("title", "")
                        // Strip HTML tags from WHO search titles
                        val title = titleRaw.replace(Regex("<[^>]*>"), "")
                        val code = entity.optString("theCode", "N/A")
                        val uri = entity.optString("id", "")
                        results.add(IcdSearchResult(code = code, title = title, uri = uri))
                    }
                }
                return@withContext results
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error in ICD-11 search, falling back.", e)
            return@withContext getLocalFallbackResults(query)
        }
    }

    private fun getLocalFallbackResults(query: String): List<IcdSearchResult> {
        val lower = query.lowercase().trim()
        val allLocalData = listOf(
            IcdSearchResult("6A70", "Single episode depressive disorder", "https://id.who.int/icd/entity/1410143890"),
            IcdSearchResult("6A71", "Recurrent depressive disorder", "https://id.who.int/icd/entity/281313880"),
            IcdSearchResult("6A72", "Dysthymic disorder", "https://id.who.int/icd/entity/46387063"),
            IcdSearchResult("6B00", "Generalized anxiety disorder", "https://id.who.int/icd/entity/1511210815"),
            IcdSearchResult("6B01", "Panic disorder", "https://id.who.int/icd/entity/1297593674"),
            IcdSearchResult("6B02", "Specific phobia", "https://id.who.int/icd/entity/1429994640"),
            IcdSearchResult("6B03", "Social anxiety disorder", "https://id.who.int/icd/entity/1897368560"),
            IcdSearchResult("6B05", "Agoraphobia", "https://id.who.int/icd/entity/240578619"),
            IcdSearchResult("6B40", "Post-traumatic stress disorder (PTSD)", "https://id.who.int/icd/entity/2070316417"),
            IcdSearchResult("6B41", "Complex post-traumatic stress disorder (Complex PTSD)", "https://id.who.int/icd/entity/585833559"),
            IcdSearchResult("6B42", "Adjustment disorder", "https://id.who.int/icd/entity/155985011"),
            IcdSearchResult("6A05", "Attention deficit hyperactivity disorder (ADHD)", "https://id.who.int/icd/entity/374567562"),
            IcdSearchResult("6A02", "Autism spectrum disorder", "https://id.who.int/icd/entity/1244592887"),
            IcdSearchResult("6A60", "Bipolar I disorder", "https://id.who.int/icd/entity/1930263630"),
            IcdSearchResult("6A61", "Bipolar II disorder", "https://id.who.int/icd/entity/877409240"),
            IcdSearchResult("6A62", "Cyclothymic disorder", "https://id.who.int/icd/entity/2130767228"),
            IcdSearchResult("6D10", "Personality disorder", "https://id.who.int/icd/entity/1922987179"),
            IcdSearchResult("6D11.5", "Borderline pattern personality specifier", "https://id.who.int/icd/entity/713337965"),
            IcdSearchResult("6B20", "Obsessive-compulsive disorder (OCD)", "https://id.who.int/icd/entity/1018318894"),
            IcdSearchResult("6B21", "Body dysmorphic disorder", "https://id.who.int/icd/entity/1769854497"),
            IcdSearchResult("6B80", "Anorexia nervosa", "https://id.who.int/icd/entity/263852482"),
            IcdSearchResult("6B81", "Bulimia nervosa", "https://id.who.int/icd/entity/1120014902"),
            IcdSearchResult("6B82", "Binge eating disorder", "https://id.who.int/icd/entity/1959728221")
        )

        if (lower.isEmpty()) return allLocalData

        return allLocalData.filter {
            it.title.lowercase().contains(lower) || it.code.lowercase().contains(lower)
        }
    }
}

File: ios/PsyPyrus/Services/IcdService.swift
import Foundation

#if os(iOS)
import UIKit
#elseif os(macOS)
import AppKit
#endif

public class IcdService {
    public static let shared = IcdService()
    private init() {}
    
    private let TAG = "IcdService" // Swift uses normal variables
    private let tokenUrl = "https://icdaccessmanagement.who.int/connect/token"
    private let searchUrl = "https://id.who.int/icd/entity/search"
    
    public var customClientId: String = ""
    public var customClientSecret: String = ""
    
    private var cachedToken: String = ""
    private var tokenExpiryTime: Double = 0
    
    private func getAccessToken() async -> String? {
        guard !customClientId.isEmpty && !customClientSecret.isEmpty else {
            return nil
        }
        
        let currentTime = Date().timeIntervalSince1970 * 1000
        if !cachedToken.isEmpty && currentTime < tokenExpiryTime {
            return cachedToken
        }
        
        guard let url = URL(string: tokenUrl) else {
            return nil
        }
        
        let credentials = "\(customClientId):\(customClientSecret)"
        guard let credentialData = credentials.data(using: .utf8) else {
            return nil
        }
        let base64Credentials = credentialData.base64EncodedString()
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Basic \(base64Credentials)", forHTTPHeaderField: "Authorization")
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        
        var components = URLComponents()
        components.queryItems = [
            URLQueryItem(name: "grant_type", value: "client_credentials"),
            URLQueryItem(name: "scope", value: "icdapi_access")
        ]
        request.httpBody = components.query?.data(using: .utf8)
        
        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
                return nil
            }
            
            if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
               let token = json["access_token"] as? String,
               let expiresIn = json["expires_in"] as? Double {
                self.cachedToken = token
                self.tokenExpiryTime = (Date().timeIntervalSince1970 * 1000) + (expiresIn - 60) * 1000
                return token
            }
            return nil
        } catch {
            print("Error getting access token in Swift: \(error.localizedDescription)")
            return nil
        }
    }
    
    public func searchIcd11(query: String) async -> [IcdSearchResult] {
        guard let token = await getAccessToken() else {
            print("Using local fallback ICD-11 search for query: \(query)")
            return getLocalFallbackResults(query: query)
        }
        
        guard let encodedQuery = query.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed),
              let url = URL(string: "\(searchUrl)?q=\(encodedQuery)") else {
            return getLocalFallbackResults(query: query)
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("v2", forHTTPHeaderField: "API-Version")
        request.setValue("en", forHTTPHeaderField: "Accept-Language")
        request.timeoutInterval = 15
        
        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
                print("Search request failed. Falling back to local data.")
                return getLocalFallbackResults(query: query)
            }
            
            var results: [IcdSearchResult] = []
            if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
               let destinationEntities = json["destinationEntities"] as? [[String: Any]] {
                for entity in destinationEntities {
                    let titleRaw = entity["title"] as? String ?? ""
                    // Strip HTML tags using simple regex
                    let title = titleRaw.replacingOccurrences(of: "<[^>]*>", with: "", options: .regularExpression)
                    let code = entity["theCode"] as? String ?? "N/A"
                    let uri = entity["id"] as? String ?? ""
                    results.append(IcdSearchResult(code: code, title: title, uri: uri))
                }
            }
            return results;
        } catch {
            print("Error in ICD-11 search in Swift, falling back: \(error.localizedDescription)")
            return getLocalFallbackResults(query: query)
        }
    }
    
    private func getLocalFallbackResults(query: String) -> [IcdSearchResult] {
        let lower = query.lowercased().trimmingCharacters(in: .whitespacesAndNewlines)
        let allLocalData = [
            IcdSearchResult(code: "6A70", title: "Single episode depressive disorder", uri: "https://id.who.int/icd/entity/1410143890"),
            IcdSearchResult(code: "6A71", title: "Recurrent depressive disorder", uri: "https://id.who.int/icd/entity/281313880"),
            IcdSearchResult(code: "6A72", title: "Dysthymic disorder", uri: "https://id.who.int/icd/entity/46387063"),
            IcdSearchResult(code: "6B00", title: "Generalized anxiety disorder", uri: "https://id.who.int/icd/entity/1511210815"),
            IcdSearchResult(code: "6B01", title: "Panic disorder", uri: "https://id.who.int/icd/entity/1297593674"),
            IcdSearchResult(code: "6B02", title: "Specific phobia", uri: "https://id.who.int/icd/entity/1429994640"),
            IcdSearchResult(code: "6B03", title: "Social anxiety disorder", uri: "https://id.who.int/icd/entity/1897368560"),
            IcdSearchResult(code: "6B05", title: "Agoraphobia", uri: "https://id.who.int/icd/entity/240578619"),
            IcdSearchResult(code: "6B40", title: "Post-traumatic stress disorder (PTSD)", uri: "https://id.who.int/icd/entity/2070316417"),
            IcdSearchResult(code: "6B41", title: "Complex post-traumatic stress disorder (Complex PTSD)", uri: "https://id.who.int/icd/entity/585833559"),
            IcdSearchResult(code: "6B42", title: "Adjustment disorder", uri: "https://id.who.int/icd/entity/155985011"),
            IcdSearchResult(code: "6A05", title: "Attention deficit hyperactivity disorder (ADHD)", uri: "https://id.who.int/icd/entity/374567562"),
            IcdSearchResult(code: "6A02", title: "Autism spectrum disorder", uri: "https://id.who.int/icd/entity/1244592887"),
            IcdSearchResult(code: "6A60", title: "Bipolar I disorder", uri: "https://id.who.int/icd/entity/1930263630"),
            IcdSearchResult(code: "6A61", title: "Bipolar II disorder", uri: "https://id.who.int/icd/entity/877409240"),
            IcdSearchResult(code: "6A62", title: "Cyclothymic disorder", uri: "https://id.who.int/icd/entity/2130767228"),
            IcdSearchResult(code: "6D10", title: "Personality disorder", uri: "https://id.who.int/icd/entity/1922987179"),
            IcdSearchResult(code: "6D11.5", title: "Borderline pattern personality specifier", uri: "https://id.who.int/icd/entity/713337965"),
            IcdSearchResult(code: "6B20", title: "Obsessive-compulsive disorder (OCD)", uri: "https://id.who.int/icd/entity/1018318894"),
            IcdSearchResult(code: "6B21", title: "Body dysmorphic disorder", uri: "https://id.who.int/icd/entity/1769854497"),
            IcdSearchResult(code: "6B80", title: "Anorexia nervosa", uri: "https://id.who.int/icd/entity/263852482"),
            IcdSearchResult(code: "6B81", title: "Bulimia nervosa", uri: "https://id.who.int/icd/entity/1120014902"),
            IcdSearchResult(code: "6B82", title: "Binge eating disorder", uri: "https://id.who.int/icd/entity/1959728221")
        ]
        
        if lower.isEmpty {
            return allLocalData
        }
        
        return allLocalData.filter {
            $0.title.lowercased().contains(lower) || $0.code.lowercased().contains(lower)
        }
    }
}
